### PR TITLE
feat(mechanics): Show stored outfits count by planet

### PR DESCRIPTION
**Pull request by jostephd**
**Feature:** This PR implements the feature request detailed and discussed in https://github.com/endless-sky/endless-sky/issues/5939#issuecomment-834559751

## Feature Details
Add a way to show the number of stored outfits by planet. In master, the map panel shows the number of stored outfits _by system_ (and #5941 doesn't change this), so there's no way to see which planets in a given system store outfits, or how many, without landing on them again.

## UI Screenshots
2021-05-07-171702_208x143_scrot

## Testing Done
Went to a system with two planets with outfitters, stored some outfits on each, took the screenshot.

(edit: Also test the case of having no flagship: https://github.com/endless-sky/endless-sky/pull/5942#pullrequestreview-668208190)

## Performance Impact
N/A

### CODE REVIEW

**Comment by petervdmeer** [comment]
Instead of this change I prefer to see the Annoyance-avoidance feature implemented (that allows the player to access all stored outfits in a system from all planets in that system, similar to what is done for ships). And if we go that route, then I feel that we wouldn't need the change from this PR since it would not matter to players which planet the outfits are stored on, as long as they know which system they need to go to.

**Comment by MasterOfGrey** [comment]
> that allows the player to access all stored outfits in a system from all planets in that system, similar to what is done for ships

Wait, you could re-outfit a ship on a different planet in the same system? That sounds hella hacky

**Comment by petervdmeer** [comment]
> Wait, you could re-outfit a ship on a different planet in the same system?

Yes, that is possible.

> That sounds hella hacky

Code-wise it's just filtering the total ships list by `planet->System` instead of filtering on just `planet` when outfitting. Other parts of the code seem more hacky than this.

**Comment by vitalchip** [comment]
Just for balance, I find the fact that you *can* outfit ships on other in system planets to be (mildly) annoying. If the flagship is on planet A and an escort on B it should not be possible to outfit/sell/change flagship.

**Comment by jostephd** [comment]
> Instead of this change I prefer to see the Annoyance-avoidance feature implemented (that allows the player to access all stored outfits in a system from all planets in that system, similar to what is done for ships). And if we go that route, then I feel that we wouldn't need the change from this PR since it would not matter to players which planet the outfits are stored on, as long as they know which system they need to go to.

I wouldn't oppose this, but so long as the annoyance-avoided feature hasn't been implemented, I think this change is a reasonable stopgap. If it does get implemented (jury seems to be out on this, over on #5210 for ships?), I wouldn't be offended if this PR is reverted outright.

**Comment by MasterOfGrey** [comment]
> Code-wise it's just filtering the total ships list by `planet->System` instead of filtering on just `planet` when outfitting. Other parts of the code seem more hacky than this.

To be clear, I meant this as being hacky in a gameplay sense, not in terms of the actual code itself.

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp]: Done in e4caa9aca02796801e43968b1f813b6a54eb28b9.

File: source/MapPanel.cpp
Comment (jostephd) [inline source/MapPanel.cpp:1189]: `escortSystems` only gets populated when the `Show escort systems on map` preference is set. Thus, the extra information about outfits is currently conditional on that preference, too. I'm not sure what the best way to fix this is.

**Comment by petervdmeer** [comment]
> I wouldn't oppose this, but so long as the annoyance-avoidance feature hasn't been implemented, I think this change is a reasonable stopgap. If it does get implemented (jury seems to be out on this, over on #5210 for ships?), I wouldn't be offended if this PR is reverted outright.

Good to hear. I hope you also wouldn't be offended if the annoyance-avoidance feature gets implemented before this gets merged. (#5961)
(And #5961 is also not merged yet, so the devs might still choose to put this PR in 0.9.14 as stopgap and keep the annoyance-avoidance feature for 0.9.15.)

**Comment by jostephd** [comment]
> Good to hear. I hope you also wouldn't be offended if the annoyance-avoidance feature gets implemented before this gets merged. (#5961)

Not in the least! If this PR becomes unnecessary, just close it. It won't be stepping on my toes or anything.

File: source/MapPanel.cpp
Comment (tehhowch) [inline source/MapPanel.cpp:69]: I think it is possibly more useful to display a tool-tip in the orbits scene, when the player hovers over the individual planet, rather than further expand the system-level tool-tip with planet-specific information.

File: source/MapPanel.h
Comment (tehhowch) [inline source/MapPanel.h:128]: If we keep this tooltip at the system level, I think it makes sense to introduce a nested class, e.g.

```c++
class SystemTooltipData {
public:
  unsigned parkedShips = 0;
  unsigned activeShips = 0;
  std::map<const Planet *, unsigned> outfits;
}
```
Then this field is just `std::map<const System *, SystemTooltipData>` instead of some unwieldy mess of `.first`s and `.second`s that is only comprehensible inside an IDE.

File: source/MapPanel.cpp
Comment (MasterOfGrey) [reply source/MapPanel.cpp:69]: Gross

Comment (jostephd) [reply source/MapPanel.cpp:69]: @tehhowch The orbits scene also shows unlandable planets, and isn't currently used for anything, so players might not realize it has a tooltip with useful information. In contrast, the system's tooltip is already used, and describes a clickable thing that even has graphical annotations that indicate when the tooltip will contain specific information. So I think putting the tooltip on the system is more useful than putting it on the orbits.

Comment (jostephd) [reply source/MapPanel.cpp:69]: @MasterOfGrey What is and why?

File: source/MapPanel.h
Comment (jostephd) [reply source/MapPanel.h:128]: As part of this PR, or as a separate PR?

File: source/MapPanel.cpp
Comment (petervdmeer) [reply source/MapPanel.cpp:69]: I like tehhowch idea. The argument that the orbits scene is not actively used is actually a good argument to start using it for more information like this.
We could also put a special marking on planets in the orbits scene that have stored items to make them more explicit/discoverable; maybe the same marker as [on systems](https://github.com/endless-sky/endless-sky/pull/4443#issuecomment-583619017)?

Comment (petervdmeer) [reply source/MapPanel.cpp:69]: For systems I would keep the total number of outfits stored in the pop-up, or maybe a listing of stored outfits per type (and then use tehhowch idea to give the per planet information).

Comment (jostephd) [reply source/MapPanel.cpp:69]: > The argument that the orbits scene is not actively used is actually a good argument to start using it for more information like this.

:thinking:

For discoverability, perhaps we could update [the help.txt "UI tutorial" messages](https://github.com/endless-sky/endless-sky/blob/13d22b84a8bd01070ab2705144511c7c69129c1f/data/help.txt#L69-L82) to point out the existence of orbits scene tooltips.

>  We could also put a special marking on planets in the orbits scene that have stored items to make them more explicit/discoverable;

Wouldn't this require a few more additions to the orbits scene, though? For starters, adding something to distinguish landable objects from other objects?

> maybe the same marker as [on systems](https://github.com/endless-sky/endless-sky/pull/4443#issuecomment-583619017)?

Should the "This system is a mission destination/stopover" markers also be applied to individual planets in the orbits scene, then? Especially in case you have missions to, say, both Luna and Earth? (And for that matter, it might be nice to have a subtle indication in the _system's_ mission destination markers that the destinations are not all on the same planet, but that's getting off-topic.)

Sorry if I sound like I have a mild case of featuritis. I'm just trying to run with @tehhowch's idea and see where it leads. If I had featuritis, I'd be proposing to use the orbits scene to indicate the government of each planet (#3768) :) 

> For systems I would keep the total number of outfits stored in the pop-up, or maybe a listing of stored outfits per type (and then use tehhowch idea to give the per planet information).

How about just the number of outfits and their total market value? Though there's a small can of worms here regarding _which_ value to show: the "As new" price, the depreciated-as-of-today sale price, or the depreciated-as-of-reaching-that-system sale price.

Comment (petervdmeer) [reply source/MapPanel.cpp:69]: The habitable (~= landable) planets already seem to stand-out with a blue color in the orbits scene.
Making the landable objects in the orbits scene match the "legend" selection (like showing friendly/hostile and showing governments when selected) sounds nice, but for a separate PR.
Adding the "mission destination/stopover" markers also to the orbit scene also sounds nice, but also for a separate PR.
Adding some tooltips could also be nice.

I would expect that storing outfits would typically be done to re-use them later. So when looking for stored outfits, then the total number and the exact types are interesting to a player. The value of the outfit(s) does not sound like something useful to display for stored outfits (especially since you can already lookup the "new price" value in the outfitter screen.)

Comment (jostephd) [reply source/MapPanel.cpp:69]: Mostly agreed, just one thing:

> I would expect that storing outfits would typically be done to re-use them later. So when looking for stored outfits, then the total number and the exact types are interesting to a player. The value of the outfit(s) does not sound like something useful to display for stored outfits (especially since you can already lookup the "new price" value in the outfitter screen.)

I proposed the number and total value of outfits as the tooltip for the system; just a summary so the tooltip doesn't get too crowded. The tooltip of the planet could give the exact list of outfits, even broken out by category (same categories as in the outfitter screen). Though it might be better to instead add a filter to the map outfitter view that shows stored outfits (like #5941 does, but without showing outfits for sale in the list at the same time; i.e., a filter that shows _only_ semidim items, rather than both semidim and dim items).

File: source/MapPanel.h
Comment (tehhowch) [reply source/MapPanel.h:128]: It would make sense to add a `SystemTooltipData` class when adding a `PlanetTooltipData` class. I think we all can agree that a map of nested pairs is not readable without IDE features.

File: source/MapPanel.cpp
Comment (petervdmeer) [reply source/MapPanel.cpp:69]: Sounds good, but I still feel that there is little value in showing the outfits total value here. Showing the total amount per planet (as done earlier in this thread) per system feels more valuable to me and there will usually not be more than 2 outfitters in a system anyway, so the risk of overcrowding also seems small.

File: source/MapPanel.h
Comment (petervdmeer) [reply source/MapPanel.h:128]: Yes, the nested classes look a lot more readable. The question from jostephd is however not yet answered, so hereby my view on that;
The readability of the affected code was not good before thisPR, but also gets a little worse by the change in this PR, so the addition of the nested classes should be done in a separate PR that gets merged earlier than this PR (which would give you a merge-conflict for this PR), or that should be addressed inside this PR (which would make this PR a little bit bigger).
I feel that the change is not too large, so combining with this PR is probably preferred from an Changelog/review perspective (but if you feel that the change is sufficiently large to deserve it's own PR, then please feel free to choose a separate PR instead.)

File: source/MapPanel.cpp
Comment (petervdmeer) [reply source/MapPanel.cpp:1188]: Is rename to `Show escorts and stored outfits on map` an option?

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp:69]: OK, so let's the keep the system tooltip as it stands, then. I'll look into implementing a per-planet tooltip, unless y'all prefer to merge the systems tooltip first and do the planets tooltip in a follow-up PR. (That might make sense if there's a release around the corner.)

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp:1188]: Combining escorts and stored outfits into a single setting sounds good; however:

- That string would overflow the space allotted to it in the preferences screen.

- If we rename the option, we'll need to teach the code to read the old name from existing preferences.txt files for backward compatibility. We might even continue to _write_ it, too, if we want to support downgrading to older versions of endless-sky without losing a change to the setting. (In theory, we could make this option the only one whose UI name is different to its preferences.txt name, but I don't think we want to do that.)

- In the longer term, introducing an abstraction between the UI name and the preferences.txt name of a setting wouldn't be the worst idea.

File: source/MapPanel.h
Comment (jostephd) [reply source/MapPanel.h:128]: Thanks, @petervdmeer. From a review perspective, I think the nested class refactoring should be a separate commit _after_ merging to master, which means it should be a separate PR, because this project squashes PRs when merging them. So I'll do a separate PR. But I'd prefer to do it as a _follow-up_ PR so I don't have to resolve conflicts, please. (I'm perfectly capable of rebasing and resolving conflicts, but it takes time, which I don't have an abundance of.)

Or if you prefer, I can squash this PR to two commits (one for the tooltips changes and one for the pairs/class refactoring) so it can be merged by rebasing rather than by squashing.

Comment (petervdmeer) [reply source/MapPanel.h:128]: Good considerations, but please keep in mind that me and most other reviewers also don't have an abundance of time available. If your approach would result in a PR that is still efficiently reviewable, then I don't see much issues with your approach. But if your approach would make this PR (more) difficult to review, then that could become a bottleneck to get this PR towards merging.

I don't have permissions to merge PRs on this repository, but I feel that you are better of creating 2 separate PRs if you feel strongly about the commits needing to be visible separately on master. (Rebase merges will take a little bit more effort from the devs and the devs already seem to be quite busy.)

Comment (tehhowch) [reply source/MapPanel.h:128]: It makes sense to me to implement both tooltip classes in the same PR. That said, if implementing orbits-based tooltips wouldn't be touching any of the same code, splitting it could make sense as there won't be much in the way of conflicts.

W.r.t. dealing with squash merges of a precursor feature branch, I've utilized [`git post-squash`](https://github.com/nomeata/git-post-squash) at work a few times to good effect.

File: source/MapPanel.cpp
Comment (MasterOfGrey) [reply source/MapPanel.cpp:69]: @jostephd sorry, my comment had been at tehhowch's suggestion of a tool-tip in the orbits scene

File: source/MapPanel.h
Comment (jostephd) [reply source/MapPanel.h:128]: > Good considerations, but please keep in mind that me and most other reviewers also don't have an abundance of time available.

Of course. I am not trying to save time for myself at the reviewers' expense. On the contrary, I proposed what, in my opinion, would _save_ time for reviewers.

> If your approach would result in a PR that is still efficiently reviewable, then I don't see much issues with your approach. But if your approach would make this PR (more) difficult to review, then that could become a bottleneck to get this PR towards merging.

Personally, I think putting the refactoring and the tooltip-adding in the same PR, and then squash-merging it, would make the resulting commit on master more difficult to review than if the two parts were separate commits on master. That's why I proposed what I did. However, it's not my call. If y'all prefer that both changes be made in the same PR, I can do that.

> It makes sense to me to implement both tooltip classes in the same PR. That said, if implementing orbits-based tooltips wouldn't be touching any of the same code, splitting it could make sense as there won't be much in the way of conflicts.

Implementing everything in the same PR feels to me a little like making the perfect the enemy of the good. Consider, there are several pieces to this puzzle:

1. Convert the nested pairs to a class.

2. Add systems tooltips.

3. Add orbits tooltips.

This PR does (2), and that part is basically ready (except that https://github.com/endless-sky/endless-sky/pull/5942#pullrequestreview-659178759 is outstanding). This PR doesn't add the new feature in (3), and neither improves on the existing technical debt in (1) nor regresses it. So if it were up to me, I'd have considered merging this PR as soon as https://github.com/endless-sky/endless-sky/pull/5942#pullrequestreview-659178759 were resolved. I mean, it's not like adding the systems tooltip before the other parts has any obvious downside, not that I can see, anyway, regardless of whether orbits tooltips do or don't touch the same code as the other parts.

Anyway, I've made my point. Please let me know how you'd like to proceed, whether I should add (1) and (3) to this PR, or whether this PR can be merged first and (1) and (3) done subsequently.

> W.r.t. dealing with squash merges of a precursor feature branch, I've utilized [`git post-squash`](https://github.com/nomeata/git-post-squash) at work a few times to good effect.

Thanks, but it sounds like a plain `git rebase --onto=foo bar baz` could do it just as well.

Comment (tehhowch) [reply source/MapPanel.h:128]: We're in a code freeze currently, waiting for a Steam release. No timeline on when it will lift, as MZ has time for the release whenever he has time for the release.

So if one approach lets you use your current free time to good effect, that's another argument for it.

Comment (petervdmeer) [reply source/MapPanel.h:128]: @jostephd You seem to have thought about this well. I also made my point and I don't have a strong preference, please feel free to choose the approach you feel is best.

Comment (jostephd) [reply source/MapPanel.h]: @tehhowch @petervdmeer Alright. I've added the refactoring to this PR, and partially squashed this PR so the refactoring is a separate commit from the tooltip adding. Thanks for the constructive discussion.

Next on my list is to figure out the preferences issue (https://github.com/endless-sky/endless-sky/pull/5942#pullrequestreview-659178759).

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp]: So, the preferences screen looks like this:

2021-05-19-002708_261x249_scrot

The option names don't currently have tooltips.

The option names must fit on a single line. (They're drawn with Table::Draw(std::string&), which renders the string on a single line.)

This being the case, I think it'll be easiest to add a separate `Show stored outfits on map` preference. Renaming the preference, aside from the backward compatibility preferences.txt reading consideration, would require teaching `Table` to have multiline cells. That seems like it would be a lot of effort for little gain, or possibly even a lot of effort for _negative_ gain, should support for multiline cells make it more difficult to extend Table in other ways in the future.

Comment (jostephd) [reply source/MapPanel.cpp]: > This being the case, I think it'll be easiest to add a separate `Show stored outfits on map` preference.

Done and force pushed. Should hopefully be good to merge now? The orbits tooltips needn't block adding the systems tooltip, I think.

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp]: > I think it is possibly more useful to display a tool-tip in the orbits scene, when the player hovers over the individual planet, rather than further expand the system-level tool-tip with planet-specific information.

Looking into this. The system tooltips are handled by MapPanel::Hover and MapPanel::DrawTooltips. The former is already marked `override` to enable MapSalesPanel::Hover. I guess I should add a MapDetailPanel::Hover, then. What should it do if the x,y coordinates are outside the orbits scene? Should it just `return false` so UI::Handle can loop on, or should it call MapPanel::Hover as MapSalesPanel::Hover does?

**Comment by jostephd** [comment]
Summary: systems tooltips are implemented and (IMHO) ready to merge; looking into implementing per-planet tooltips in the orbits scene (https://github.com/endless-sky/endless-sky/pull/5942#discussion_r637606361).

File: source/MapPanel.cpp
Comment (petervdmeer) [inline source/MapPanel.cpp:1199]: Not a bug you introduced, but I just realized that this can make activeShips negative if the player has no flagship. (And I just tested this, it shows -1 escorts if I open the map after I just sold my flagship with the original implementation.)

File: source/MapPanel.cpp
Comment (petervdmeer) [inline source/MapPanel.cpp:1224]: Maybe also add an OR-condition check if t.outfits size is larger than 1 (in case one of the planets becomes un-landable but still has outfits).

**Review by petervdmeer — APPROVED** [review]
Looks pretty good. I have 2 minor review-comments, but I don't consider them sufficiently critical not to approve.

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp]: Good catch. Added without testing.

**Comment by jostephd** [comment]
Both comments addressed, I think. Thanks again @petervdmeer for reviewing.

File: source/MapPanel.cpp
Comment (tehhowch) [inline source/MapPanel.cpp:1227]: This is pretty much the only part of the PR I don't particularly like, since it is putting the planet-specific data onto the system tooltip.

That said, it won't cause any crazy-looking tooltips in vanilla, since it requires X outfitters in the system to get X new lines (and we have much ghastlier UI issues with 3+ _landable_ objects).

I'm fine with this as a stopgap, but do want to see the per-planet system-level callout get removed in the follow up that augments the orbits UI.

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp]: `const` because I always use that for variables that aren't written to. `&` because without that it always looks to me like it does a copy, but I'll drop it.

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp]: Added

File: source/MapPanel.cpp
Comment (jostephd) [reply source/MapPanel.cpp]: Changed

**Comment by jostephd** [comment]
Implemented some of the suggestions. The rest I don't have time to do right now, I'll do them later and update this PR.

File: source/MapPanel.cpp
Comment (petervdmeer) [reply source/MapPanel.cpp]: When this becomes an issue, then we can always change the tooltip to "You have <nr> outfits stored over <nr> planets in this system."
The issue is for now not very likely, because it will only be an issue when there are many outfitters in a system, and when the player decides to store outfits on many places.

File: source/MapPanel.cpp
Comment (petervdmeer) [reply source/MapPanel.cpp]: I uploaded a fix for the double count (that also caused the comparison to be removed).

File: source/MapPanel.cpp
Comment (petervdmeer) [reply source/MapPanel.cpp]: This code got replaced.

**Comment by petervdmeer** [comment]
Merged master and processed open review comments.

File: source/MapPanel.cpp
Comment (tehhowch) [inline source/MapPanel.cpp:112]:
```suggestion
		for(auto &&stellarObject : system.Objects())
```
No need to make a copy in order to have a mutable reference.

Comment (quyykk) [reply source/MapPanel.cpp:112]: I don't see the copy here, both lines are in fact functionally identical. Am I missing something?

**Pull request merged**
